src/util/Timer.cpp

```diff
@@ -281,27 +281,26 @@ VirtualClock::crank(bool block)
 
     nWorkDone -= nRealTimerCancelEvents;
 
+    if (mMode == VIRTUAL_TIME && nWorkDone == 0)
+    {
+        // If we did nothing and we're in virtual mode,
+        // we're idle and can skip time forward.
+        nWorkDone += advanceToNext();
+    }
+
     {
         std::lock_guard<std::mutex> lock(mDelayedExecutionQueueMutex);
         if (!mDelayedExecutionQueue.empty())
         {
             block = false;
-            // count this as work
-            nWorkDone += 1;
         }
         for (auto&& f : mDelayedExecutionQueue)
         {
             mIOService.post(std::move(f));
         }
         mDelayedExecutionQueue.clear();
     }
-
-    if (mMode == VIRTUAL_TIME && nWorkDone == 0)
-    {
-        // If we did nothing and we're in virtual mode,
-        // we're idle and can skip time forward.
-        nWorkDone += advanceToNext();
-    }
+    mDelayExecution = false;
```

Comment:
I didn't want to use mutexes/atomic for mDelayExecution, because I don't think it is worth it.

In some cases of race conditions around the mDelayExecution member it is possible that job will be posted directly to io_service instead of to the delayed queue and in some cases it will be the opposite, but that should not matter much.

I do not understand what are you saying about queuing 2 events in a row. Which execution will make that happen?